### ISSUE TITLE
Paper #24: A Principled Approach to GraphQL Query Cost Analysis

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#68]
This issue is related to PR #55 

Authors are applying for the *reusable*, *functional*, and *available* badges.

Author contact: @Alan-Cha

**Comment by kstolee** [comment]
# Authors

| Name | Email | GitHub|
|------|-------|-------|
| Alan Cha (corresponding author) | Alan.Cha1@ibm.com | https://github.com/Alan-Cha |
| Erik Wittern | erik.wittern@ibm.com | https://github.com/ErikWittern |
| Guillaume Baudart | guillaume.baudart@ibm.com | https://github.com/gbdrt |
| James C. Davis | davisjam@vt.edu | https://github.com/davisjam |
| Louis Mandel | lmandel@us.ibm.com | https://github.com/mandel |
| Jim A. Laredo | laredoj@us.ibm.com | https://github.com/laredo |

**Comment by art-reviewer** [comment]
[Paper](https://github.com/Alan-Cha/fse20/blob/38cef7ada1dde63ccfcd679b242b10ba06819e48/submissions/reusable/FSE-24/graphql-paper.pdf)

**Comment by anon-eval** [comment]
@Alan-Cha 
Thank you for submitting the artifacts. Congratulations on the paper :)

**TLDR**: Based on the current version of your submission, I recommend the **available ** and **functional** and the **reusable** badge. That said, I did have some trouble in the installation and it would be great if you could update `INSTALL.md` with this information.

1. Please add a link to your github repository since you refer `folders` and `paths` from that repo..
2. Please add a requirements.txt file with the python libraries needed and the versions.

I arrived at this conclusion using the following metrics
1. **Installation** 🟡 : The installation runs successfully but I needed to figure out the dependencies from the errors.
2. **Public Availability**✅ : The tool is publicly available on zenodo and github with a DOI.
3. **Documentation** ✅  : The tool is well documented for reproducing the results and executing it.
4. **Consistent/Complete**✅ : The results are consistent on multiple executions and the results in the paper can be reproduced using the scripts provided.
5. **Exercisable** ✅ : The tool is exercisable on the datasets provided in the repository
6. **Evidence of verification and validation** ✅ : I was able to only validate the results for github and yelp datasets.
7 **Reusability** ✅ : The authors clearly document how the tool can be run on custom github and yelp queries.

Thanks :)

**Comment by hopefully-anon** [comment]
@Alan-Cha 

Until the installation dependencies are identified and updated, I can only recommend the **available** badge.   Further, could you clarify what OS and `npm` version this should work for?  

Specifically, am struggling to get **_verification_** running on a clean copy of **Ubuntu 18.04** after upgrading `npm` to version 6.14.5


@anon-eval, if you have the missing dependencies for FSE-24 handy, could you list them?

**Comment by AnonReviewer626** [comment]
@hopefully-anon Can you provide the error messages to guide @Alan-Cha?

**Comment by anon-eval** [comment]
@hopefully-anon 
Here are the dependencies I needed to install
```
numpy==1.16.6
matplotlib==3.2.1
```
Please note that the latest version of `numpy(v1.18.5)` did not work.

**Comment by Alan-Cha** [comment]
@anon-eval @hopefully-anon Thank you so much for your feedback! I will reference the GitHub repo and add the dependencies to the `INSTALL.md`. 

@hopefully-anon If you could provide the error messages, that would be very helpful!

**Comment by hopefully-anon** [comment]
Perhaps I'm missing steps from the `README`?

Starting with a new AWS Ubuntu 18.04 instance:

## Installation:  upgrade `npm` 

`sudo apt-get install npm`

`npm --version` yields `3.5.2`

This fails `npm ci` so I upgraded `npm`  as per the instructions here:
`https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-18-04`

Now `npm --version` yields `6.14.4`, so `npm ci` works.


## Installation:  Missing from `README` Yelp Developer Beta

Running `npm run verify` fails:

 and upon investigation the yelp api is yielding:
`{"error": {"code": "NOT_IN_DEVELOPER_BETA", "description": "This endpoint requires you be in our developer beta. To join, visit your \"Manage App\" page and click the join beta button."}}`

Joined the `Yelp` developer beta program.

## Current challenge:

`npm run verify` fails:

```
src/get-github-schema.ts:1:24 - error TS2307: Cannot find module 'node-fetch' or its corresponding type declarations.

1 import * as fetch from "node-fetch";
                         ~~~~~~~~~~~~

src/get-github-schema.ts:2:21 - error TS2307: Cannot find module 'fs' or its corresponding type declarations.

2 import * as fs from "fs";
                      ~~~~

src/get-github-schema.ts:8:10 - error TS7006: Parameter 'res' implicitly has an 'any' type.

8   .then((res) => {
           ~~~

src/get-github-schema.ts:17:10 - error TS7006: Parameter 'text' implicitly has an 'any' type.

17   .then((text) => {
            ~~~~
```

Attempted to install `node-fetch`: 

```
npm install node-fetch
```

returns:

```
+ node-fetch@2.6.0
updated 1 package and audited 11 packages in 0.35s
found 0 vulnerabilities
```

(Maybe this needs node-fetch 3.0?)

Any assistance would be much appreaciated.

**Comment by Alan-Cha** [comment]
@hopefully-anon 

You are right. We should also clarify that you need a GitHub personal access token and a Yelp API key in the `INSTALL.md`. I will look into that.

***

Did you install `node-fetch` in the `verification/` folder? If so, I might know what the problem is and it is related to the fact that we structured our source code as independent components. 

For example, `graphql-schemas/` contains the scripts needed to fetch the GitHub and Yelp schemas while `analysis-config/` contains the scripts to convert the fetched schemas into the proper configurations. Each folder is set up as an independent Node.js project. 

I think we overlooked a script in the `verification/` folder that will go into the other folders and ensure that their dependencies have been installed.

For example, to manually solve the current problem that you are facing, you would need to go into`graphql-schemas/` folder and run `npm ci`. The same goes for all the other components that the verification depends on. I will add an additional script that will do this. If you could confirm this (you should see a different set of errors from a different component), that would be helpful. 

Apologies for any confusion. While creating artifact, we worked on the verification component last and by that time, we already had the dependencies installed in the other components. It was not apparent at the time that we needed this kind of script.

**Comment by hopefully-anon** [comment]
Hi @Alan-Cha,

Yes, running `npm ci` in `graphql-schemas/` and `analysis-config/` helped overcome the import problems:

Now running `npm run verify` yields:

```
...
> verification@1.0.0 unzip-corpuses /home/ubuntu/Alan-Cha-graphql-complexity-paper-artifact-7cdd502/verification
> cd ../query-response/github && tar xvf github-query-response-part1.json.zip && tar xvf github-query-response-part2.json.zip && cd ../yelp && tar xvf yelp-query-response.json.zip && cd ..

tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! verification@1.0.0 unzip-corpuses: `cd ../query-response/github && tar xvf github-query-response-part1.json.zip && tar xvf github-query-response-part2.json.zip && cd ../yelp && tar xvf yelp-query-response.json.zip && cd ..`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the verification@1.0.0 unzip-corpuses script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by AnonReviewer626** [comment]
@Alan-Cha Would something like this work?

```
for f in `find . -name package.json`; do
  cd `dirname $f`; npm install; cd -
```

**Comment by hopefully-anon** [comment]
Installed `unzip`

```
sudo apt-get install unzip
```
and modified `verification/package.json` by replacing

```
"unzip-corpuses": "cd ../query-response/github && tar xvf github-query-response-part1.json.zip && tar xvf github-query-response-part2.json.zip && cd ../yelp && tar xvf yelp-query-response.json.zip && cd ..",
```

with

```
"unzip-corpuses": "cd ../query-response/github && unzip github-query-response-part1.json.zip && unzip github-query-response-part2.json.zip && cd ../yelp && unzip yelp-query-response.json.zip && cd ..",
```

Now `npm run verify` is running:

```
...
Processing GitHub corpus, 877 out of 5000
...
```

**Comment by hopefully-anon** [comment]
@AnonReviewer626 

I tried the `find .`  loop approach but had problems connecting to `ibm` servers:

```
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t git@github.ibm.com:graphql-research/graphql-query-generator.git
npm ERR! 
npm ERR! git@github.ibm.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR! 
npm ERR! exited with error code: 128
```

**Comment by hopefully-anon** [comment]
I  recommend the **functional** and **available** badges.    


The installation challenges seem to exceed those required to earn reusable.

Install 🟡 :   Challenging. The `README` instructions require further details about the Yelp API-Key process, an additional script is required to fulfill `npm` dependencies, and the `npm` version from a clean ubuntu 18.08 install needs to be upgraded, also had to modify some unix commands within `verfication/package.json` to get `npm run verify` to run.
Available Publicly ✅ : 
Documentation ✅ : The tool is has good documentation but is missing some details for the installation.
Consistency/Completeness ✅ : the output appears to match the contents of the paper and covers all the main points.
Code is Exercisable 🟡 : The tool is exercisable on the datasets provided in the repository ( after modifying the scripts).   Artifact required several changes to run the `verification` portion. 

Was able to successfully run the `verification` programs as expected.

**Comment by Alan-Cha** [comment]
@AnonReviewer626 @hopefully-anon Thank you for all the suggestions! @hopefully-anon Thank you also for your perseverance in running the verification with all the changes you had to make. I apologize for the inconvenience. I am going to look into all of this and make sure that it is an easy, one-step process.

***

```
for f in `find . -name package.json`; do
  cd `dirname $f`; npm install; cd -;
done;
```

Unfortunately, I do not think this is the best solution as it will also look for the `package.json` in the dependencies. No worries however, I can easily add a small script that'll install the modules in the desired components. 

***

> I tried the find . loop approach but had problems connecting to ibm servers:

This is another oversight. Sorry for that. We open-sourced this repository but it seems like we are still referencing the IBM-private repo in a few places. I will make sure that we correctly change all of these references.

**Comment by AnonReviewer626** [comment]
> Unfortunately, I do not think this is the best solution as it will also look for the package.json in the dependencies. No worries however, I can easily add a small script that'll install the modules in the desired components.

Only if the dependencies have already been partially installed ;-). However, handling this in a small script would be more appropriate.

**Comment by Alan-Cha** [comment]
@anon-eval @hopefully-anon I believe I have made all the request changes. To name some of the changes I've made:

* I described the dependencies in more detail in the `INSTALL.md`
* I added a reference to the GitHub repository
* I added a `requirements.txt` to install the Python packages in `measured-complexity/` where Python is used
* I have also added a script to install the Node modules in the other subdirectories
* I replaced references to the IBM-internal repo with the open-sourced one

Please let me know if there is anything else I missed!

**Comment by kstolee** [comment]
We seem to have agreement on ```available``` and ```functional```. 

@hopefully-anon after the updates, are you able to recommend ```reusable``` like @anon-eval ?

**Comment by a-hopefully-friendly-reviewer** [comment]
Thank you @Alan-Cha for the improvements in the _review rebuttal_ (#154), I can recommend `reusable`

**Comment by hopefully-anon** [comment]
Thank you @Alan-Cha for the improvements in the _review rebuttal_ (#154), I can recommend `reusable`

**Comment by kstolee** [comment]
Marking this evaluation as done. Congratulations @Alan-Cha !
